packages/shell-api/src/bulk.ts

```diff
@@ -0,0 +1,217 @@
+  }
+
+  update(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.update(op);
+    return this._parentBulk;
+  }
+
+  upsert(): BulkFindOp {
+    assertArgsDefined();
+    this._innerFind.upsert();
+    return this;
+  }
+}
+
+
+@shellApiClassDefault
+@hasAsyncChild
+export default class Bulk extends ShellApiClass {
+  _mongo: Mongo;
+  _collection: any; // to avoid circular ref
+  _batchCounts: any;
+  _executed: boolean;
+  _batches: any;
+  _innerBulk: any;
+
+  constructor(collection, innerBulk) {
+    super();
+    this._collection = collection;
+    this._mongo = collection._mongo;
+    this._innerBulk = innerBulk;
+    this._batches = [];
+    this._batchCounts = {
+      nInsertOps: 0,
+      nUpdateOps: 0,
+      nRemoveOps: 0
+    };
+  }
+
+  /**
+   * Internal method to determine what is printed for this class.
+   */
+  _asPrintable(): string {
+    return this.tojson();
+  }
+
+  /**
+   * Internal helper for emitting collection API call events.
+   *
+   * @param methodName
+   * @param methodArguments
+   * @private
+   */
+  private _emitBulkApiCall(methodName: string, methodArguments: Document = {}): void {
+    this._mongo._internalState.emitApiCall({
+      method: methodName,
+      class: 'Bulk',
+      db: this._collection._database._name,
+      coll: this._collection._name,
+      arguments: methodArguments
+    });
+  }
+
+  @returnsPromise
+  async execute(writeConcern?: WriteConcern): Promise<BulkWriteResult> {
+    if (this._executed) {
+      throw new MongoshInvalidInputError('A bulk operation cannot be re-executed');
+    }
+
+    if (this._innerBulk.s !== undefined && Array.isArray(this._innerBulk.s.batches)) {
+      this._batches = [...this._innerBulk.s.batches];
+      this._batches.push(this._innerBulk.s.currentBatch);
+    }
+    const result = await this._innerBulk.execute();
+    this._executed = true;
+    this._emitBulkApiCall('execute', { operations: this._batches, writeConcern: writeConcern });
+    return new BulkWriteResult(
+      !!result.result.ok, // acknowledged
+      result.result.nInserted,
+      result.result.insertedIds,
+      result.result.nMatched,
+      result.result.nModified,
+      result.result.nRemoved,
+      result.result.nUpserted,
+      result.result.upserted
+    );
+  }
+
+  find(query: Document): BulkFindOp {
+    assertArgsDefined(query);
+    return new BulkFindOp(this._innerBulk.find(query), this);
+  }
+
+  insert(document: Document): Bulk {
+    this._batchCounts.nInsertOps++;
+    assertArgsDefined(document);
+    this._innerBulk.insert(document);
+    return this;
+  }
+
+  tojson(): any {
+    const batches = this._innerBulk.s.batches.length + Number(this._innerBulk.s.currentBatch !== null);
```

Comment:
I find this concise but hiding a bit of the intent, wonder if this could be more clear?

```
let nBatches = this._innerBulk.s.batches.length;

if (this._innerBulk.s.currentBatch !== null) {
   this._innerBulk.s.batches.length += 1;
}
``` 

`this._innerBulk.s.currentBatch !== null` would return true also if `this._innerBulk.s.currentBatch` is `undefined`, is it intentional?

Reply:
Sure, I can definitely clarify the code here. If currentBatch is undefined then there's an error, so maybe we should just check that directly. Null is valid, just means there's no batch currently being created.